### PR TITLE
Feat/rpc endpoints to fetch data from key

### PR DESCRIPTION
**Pull request by hugocaillard**
### Description

Add two new endpoints:
- `/v2/clarity_marf_value/:clarity_marf_key`
- `/v2/clarity_metadata/:principal/:contract_name/:clarity_metadata_key`

I'm currently working on a Clarinet feature that allows to simulate running (or, said differently, to fork the mainnet state in the simnet data store). This requires the ability to fetch values from marf and metadata keys.

These new endpoints are similar to already existing endpoint (such as `getdatavar`,  `getmapentry`,  `getcontractsrc`, etc).

### Applicable issues

N/A

### Additional info (benefits, drawbacks, caveats)

Read more about this feature in the Clarinet issue: https://github.com/hirosystems/clarinet/issues/1503

I confirmed that the 2 new endpoints allow to achieve the desired goal by running a local devnet with these changes and forking the Clarinet simnet state from it. 

### Checklist

- [x] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)

### CODE REVIEW

File: clarity/src/vm/representations.rs [hunk 1]
```diff
@@ -51,6 +51,8 @@ lazy_static! {
         "({})|({})",
         *STANDARD_PRINCIPAL_REGEX_STRING, *CONTRACT_PRINCIPAL_REGEX_STRING
     );
+    static ref CLARITY_NAME_NO_BOUNDARIES_REGEX_STRING: String =
```
Comment (jcnelson) [inline clarity/src/vm/representations.rs:54]: This needs boundaries.  Unbounded regexes whose evaluation is accessible via the network are a DDoS bug waiting to happen.

File: stackslib/src/net/api/getclaritymarfvalue.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+/// Decode the HTTP request
+impl HttpRequest for RPCGetClarityMarfValueRequestHandler {
+    fn verb(&self) -> &'static str {
+        "GET"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(&format!(
+            r"^/v2/clarity_marf_value/(?P<clarity_marf_key>(vm-epoch::epoch-version)|({})|({}))$",
```
Comment (jcnelson) [inline stackslib/src/net/api/getclaritymarfvalue.rs:64]: I strongly disagree with using the MARF key as input.  To do so would require an explanation to an RPC API consumer of how to construct a Clarity MARF key, which essentially requires them to read the majority of [SIP-005](https://github.com/stacksgov/sips/blob/main/sips/sip-005/sip-005-blocks-and-transactions.md).  Instead, you should do the following:

* Accept the _hash_ of the key as input.  Then, you don't have to worry about verifying that it is well-formed, and you don't have to worry about explaining to the RPC API consumer how to construct it.  Furthermore, it's much safer and easier to determine if a string is a well-formed hash instead of a well-formed Clarity MARF key.
* Update the Clarity DB to take the hash as an argument, instead of the key (in support of the above).

Comment (jcnelson) [reply stackslib/src/net/api/getclaritymarfvalue.rs:64]: Also, I dislike the path you chose, since it doesn't leave room for semantic expansion.  Instead, can you use the following:  `/v2/marf/clarity/{key-hash}`, where `key-hash` is the hash of the SIP-005 key.

File: stackslib/src/net/api/getclaritymetadata.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+/// Decode the HTTP request
+impl HttpRequest for RPCGetClarityMetadataRequestHandler {
+    fn verb(&self) -> &'static str {
+        "GET"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(&format!(
+            r"^/v2/clarity_metadata/(?P<address>{})/(?P<contract>{})/(?P<clarity_metadata_key>(analysis)|({}))$",
```
Comment (jcnelson) [inline stackslib/src/net/api/getclaritymetadata.rs:64]: Please use `/v2/clarity/metadata/{principal}/{contract-name}/{clarity_metadata_key}` as your path.

File: stackslib/src/net/api/getclaritymetadata.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    fn try_handle_request(
+        &mut self,
+        preamble: HttpRequestPreamble,
+        contents: HttpRequestContents,
+        node: &mut StacksNodeState,
+    ) -> Result<(HttpResponsePreamble, HttpResponseContents), NetError> {
+        let contract_identifier = self.contract_identifier.take().ok_or(NetError::SendError(
+            "`contract_identifier` not set".to_string(),
+        ))?;
+        let clarity_metadata_key = self.clarity_metadata_key.take().ok_or(NetError::SendError(
```
Comment (jcnelson) [inline stackslib/src/net/api/getclaritymetadata.rs:120]: Clarity metadata keys have a well-defined structure, so the RPC endpoint should return HTTP 400 if the metadata key is ill-formed.  You will need to expand this method (or `try_parse_request` -- up to you) to validate the structure of the Clarity metadata key, including determining whether or not the key's `StoreType` and `var_name` values are supported values.  `var_name` will, unfortunately, take some effort because the Clarity DB codebase uses bare string literals in its calls to `ClarityDatabase::make_metadata_key()` instead of enums, so your PR should address this as well.

File: stackslib/src/net/httpcore.rs [hunk 5]
```diff
@@ -247,6 +247,17 @@ pub mod request {
         Ok(txid)
     }
 
+    /// Get a clarity key (MARF or Metadata) from a path's captures, given the name of the regex field.
+    pub fn get_clarity_key(captures: &Captures, clarity_key: &str) -> Result<String, HttpError> {
```
Comment (jcnelson) [inline stackslib/src/net/httpcore.rs:251]: Because this is specific to your handler, you should implement it there instead.

File: clarity/src/vm/representations.rs [hunk 6]
```diff
@@ -64,6 +66,19 @@ lazy_static! {
...
+        r"vm::{}::\d+::({})",
+        *CONTRACT_PRINCIPAL_REGEX_STRING,
+        *CLARITY_NAME_NO_BOUNDARIES_REGEX_STRING,
+    );
+    pub static ref MARF_KEY_FOR_QUAD_REGEX_STRING: String = format!(
+        r"{}::[0-9a-fA-F]+",
+        *MARF_KEY_FOR_TRIP_REGEX_STRING,
+    );
+    pub static ref METADATA_KEY_REGEX_STRING: String = format!(
+        r"vm-metadata::\d+::(contract|contract-size|contract-src|contract-data-size|({}))",
```
Comment (jcnelson) [inline clarity/src/vm/representations.rs:79]: You'll need boundaries here if you want this to be used in RPC calls.

File: clarity/src/vm/representations.rs [hunk 7]
```diff
@@ -64,6 +66,19 @@ lazy_static! {
         Regex::new(format!("^{}$|^__transient$", CONTRACT_NAME_REGEX_STRING.as_str()).as_str())
             .unwrap()
     };
+    pub static ref MARF_KEY_FOR_TRIP_REGEX_STRING: String = format!(
+        r"vm::{}::\d+::({})",
+        *CONTRACT_PRINCIPAL_REGEX_STRING,
+        *CLARITY_NAME_NO_BOUNDARIES_REGEX_STRING,
```
Comment (jcnelson) [inline clarity/src/vm/representations.rs:72]: You'll need boundiares here if you want this to be used in RPC calls.

File: clarity/src/vm/representations.rs (hunk 1 above)
Comment (hugocaillard) [reply clarity/src/vm/representations.rs:54]: All of the regex string I added end up being formatted together and exclusively used with boundaries ([here](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/stackslib/src/net/api/getclaritymarfvalue.rs#L63-L66) and [here](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/stackslib/src/net/api/getclaritymetadata.rs#L63-L68))

But since they are only used there, maybe I should build the regex more exclusively in these two files and not exposed in representation.rs. Would that be a better approach?

File: stackslib/src/net/api/getclaritymetadata.rs (hunk 4 above)
Comment (hugocaillard) [reply stackslib/src/net/api/getclaritymetadata.rs:120]: Apparently the metadata `var_name` can be an arbitrary string, where here it's a [variable_name](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/clarity/src/vm/database/clarity_db.rs#L1329), a [map_name](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/clarity/src/vm/database/clarity_db.rs#L1467), a [token_name](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/clarity/src/vm/database/clarity_db.rs#L1811)).

Can we really use an enum here?

File: stackslib/src/net/api/getclaritymarfvalue.rs (hunk 2 above)
Comment (hugocaillard) [reply stackslib/src/net/api/getclaritymarfvalue.rs:64]: In my use case (of Clarinet consuming this API), I already know the key (but not the hash). But I agree that it's not ideal to have the key as a use input.
Can you guide me to some code where I could see how to get the key <> key_hash conversion?

> Update the Clarity DB to take the hash as an argument, instead of the key (in support of the above).

Not sure to follow here, where should I perform this change?

--- 

I think I just lack knowledge about this repo to fully address this comment, thanks for providing extra context 🙏

Comment (jcnelson) [reply stackslib/src/net/api/getclaritymarfvalue.rs:64]: Sure -- the (string) key's bytes are hashed via SHA512/256 to produce the MARF key.

> Not sure to follow here, where should I perform this change?

The trait `ClarityBackingStore` needs to be modified to add a method like `get_data_with_proof()`, but which takes an implementation of the `MarfTrieId` trait instead of a `&str` key.  There currently isn't a way to query the MARF via the ClarityDB with a bare hash.

File: stackslib/src/net/api/getclaritymetadata.rs (hunk 4 above)
Comment (jcnelson) [reply stackslib/src/net/api/getclaritymetadata.rs:120]: Yes; however, some reserved `var_name` strings can only be paired with certain `StoreType`s.  For example, the use of `StoreType::Contract` would require `var_name` to be `contract-size`, `contract-src`, `contract-data-size`, or `contract`, but nothing else.

One way to implement these constraints could be to implement an enum to capture all of the valid `StoreType` / `var-name` pairings, and permit only the `StoreType` variants which allow arbitrary `var_name` values to have arbitrary `var_name` values.

File: stackslib/src/net/api/getclaritymarfvalue.rs (hunk 2 above)
Comment (obycode) [reply stackslib/src/net/api/getclaritymarfvalue.rs]: > I strongly disagree with using the MARF key as input. To do so would require an explanation to an RPC API consumer of how to construct a Clarity MARF key, which essentially requires them to read the majority of [SIP-005](https://github.com/stacksgov/sips/blob/main/sips/sip-005/sip-005-blocks-and-transactions.md).

Why would getting the hash of this key be any easier for a user of this API? Wouldn't that be even harder? Is there some scenario where a user would already have the hash and then would want to use this endpoint?

Note that adding a new version of `get_data_with_proof` that takes in the `TriePath` would also require some significant changes to the `MemoryBackingStore`, since it currently directly uses the key to access the database.

Comment (jcnelson) [reply stackslib/src/net/api/getclaritymarfvalue.rs]: > Why would getting the hash of this key be any easier for a user of this API? Wouldn't that be even harder? Is there some scenario where a user would already have the hash and then would want to use this endpoint?

For the client, I think both approaches are equivalently difficult.  The hard part either way is constructing the key.  The difference between sending the key versus the hash of the key is trivial.

However, when comparing the difficulty of implementing the API (including changes to the Clarity DB) to only accept a hash versus the whole metadata key, the former is actually much easier.  First, the Clarity DB already performs lookups into the MARF by hash (albeit via a few layers of indirection), so it's not a very big lift to add a trait method for loading a serialized value by `TrieHash` and adding a MARF method to lookup by `TrieHash`.  Second, by accepting a key, we would commit the API to validating that key so as to property return HTTP 400 responses [1] if the key was semantically invalid.  Adding all of the validation logic and test coverage for it would vastly exceed the work required to support lookups by `TrieHash`.

[1] I'm not a fan of returning a blanket 404 for invalid input.  A 404 specifically means that the resource could not be found, and doesn't tell the user what the actually need to know.  By contrast, any 32-byte value is a valid MARF key, so HTTP 400 would only be necessary if the value was not a hex-encoded 32-byte hash.

Comment (obycode) [reply stackslib/src/net/api/getclaritymarfvalue.rs]: Sure, I understand that argument. Beware of the changes needed to the `MemoryBackingStore` to support lookups by `TrieHash` -- this PR will unfortunately need to grow significantly. This will be more work than we would hope, but that's probably okay, since changes there are not affecting critical code.

Comment (hugocaillard) [reply stackslib/src/net/api/getclaritymarfvalue.rs]: It's a bit hard to follow but I've some questions based on my understading; 

>  "any 32-byte value is a valid MARF key, so HTTP 400 would only be necessary if the value was not a hex-encoded 32-byte hash"

We would return a 404 if the hash encodes an invalid key right? So the difference for the client seems to be quite the same to me. It's semantically sligthly more correct to return a 404 for a random 32-byte value than it is for an invalid key, but the result for the client is the same.

> "so it's not a very big lift to add a trait method for loading a serialized value by `TrieHash`"

I'm actually struggling to implement it in the MemoryBackingStore (around [here](https://github.com/stacks-network/stacks-core/blob/24566459e5a8bdb399e0fe2efbfa0dadc571c851/clarity/src/vm/database/sqlite.rs#L323-L329)), because I don't have access `TrieHash::from_data` here.
I don't see how to get from the key to the Hash.
Maybe it's trivial but I just don't know the repository good enough.

Comment (jcnelson) [reply stackslib/src/net/api/getclaritymarfvalue.rs]: > We would return a 404 if the hash encodes an invalid key right?

If this API endpoint only took a hash, then the only application responses to expect are 200 (with the serialized data) or 404 (if the hash does not map to a value).  The question of whether or not the hash corresponds to a semantically-valid but absent key or a semantically-invalid key does not need to be answered by the API endpoint.

> It's semantically sligthly more correct to return a 404 for a random 32-byte value than it is for an invalid key, but the result for the client is the same.

That's the thing -- I think it's plausible that this makes a significant difference to the client.  If the client is an application that loads Clarity data, then the difference between receiving a 400 versus a 404 is that the former requires a bugfix (the client made an invalid key) and would likely result in the user being presented with an error condition to report, whereas the latter indicates that the submitted key is valid but not mapped.  Treating both as 404's would lead to a poorer user experience, since users would receive false negatives -- a bug in the application's key construction would be treated the same as the user submitting a query for nonexistent data.

Comment (jcnelson) [reply stackslib/src/net/api/getclaritymarfvalue.rs]: I'm happy to add the requisite Clarity DB patches to this PR, once I'm done helping @obycode with https://github.com/stacks-network/stacks-core/pull/5420.

Comment (jcnelson) [reply stackslib/src/net/api/getclaritymarfvalue.rs]: I went ahead and added the requisite ClarityDB functionality, and plumbed it through into this RPC endpoint.

File: stackslib/src/net/api/getclaritymetadata.rs [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+};
+use crate::net::httpcore::{
+    request, HttpPreambleExtensions, HttpRequestContentsExtensions, RPCRequestHandler,
+    StacksHttpRequest, StacksHttpResponse,
+};
+use crate::net::{Error as NetError, StacksNodeState, TipRequest};
+
+lazy_static! {
+    static ref CLARITY_NAME_NO_BOUNDARIES_REGEX_STRING: String =
+        "[a-zA-Z]([a-zA-Z0-9]|[-_!?+<>=/*])*|[-+=/*]|[<>]=?".into();
```
Comment (jcnelson) [inline stackslib/src/net/api/getclaritymetadata.rs:38]: This regex is very problematic because it matches strings of unbounded length, which opens a DoS vector on the node.

File: stackslib/src/net/api/getclaritymetadata.rs [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+    static ref METADATA_KEY_REGEX_STRING: String = format!(
+        r"vm-metadata::\d+::(contract|contract-size|contract-src|contract-data-size|({}))",
```
Comment (jcnelson) [inline stackslib/src/net/api/getclaritymetadata.rs:40]: Same problem here with `\d+`

Comment (jcnelson) [reply stackslib/src/net/api/getclaritymetadata.rs:40]: Since the `\d+` must be only one of a handful of values, you should just match them explicitly.

**Comment by jcnelson** [comment]
I went ahead and made the requisite changes to the MARF, `ClarityBackingStore` and all its implementations, and `ClarityDB` so you can load values by `TrieHash`.  I also removed `TriePath` from the codebase, since it's used exactly the same way that `TrieHash` is.

My comments on the getter for Clarity metadata still stand, however.  If we're going to take metadata keys as input, we need to validate them and return HTTP 400 if they're semantically invalid.

File: stacks-common/src/types/chainstate.rs [hunk 10]
```diff
@@ -30,6 +30,68 @@ impl_byte_array_serde!(TrieHash);
...
+        let s = format!("{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
+                          self.0[0],     self.0[1],       self.0[2],       self.0[3],
+                          self.0[4],     self.0[5],       self.0[6],       self.0[7],
+                          self.0[8],     self.0[9],       self.0[10],      self.0[11],
+                          self.0[12],    self.0[13],      self.0[14],      self.0[15],
+                          self.0[16],    self.0[17],      self.0[18],      self.0[19],
+                          self.0[20],    self.0[21],      self.0[22],      self.0[23],
+                          self.0[24],    self.0[25],      self.0[26],      self.0[27],
+                          self.0[28],    self.0[29],      self.0[30],      self.0[31]);
+        s
```
Comment (obycode) [inline stacks-common/src/types/chainstate.rs:91]:
```rust
pub fn to_string(&self) -> String {
    self.0.iter().map(|byte| format!("{:02x}", byte)).collect()
}
```

**Comment by jcnelson** [comment]
The reason we *didn't* do that is because that invokes the allocator too
much, per perf.

File: stackslib/src/net/api/getclaritymarfvalue.rs (hunk 2 above)
Comment (hugocaillard) [reply stackslib/src/net/api/getclaritymarfvalue.rs]: Thanks a lot @jcnelson, this is super super helpful.
I'm currenlty working on addressing your other remaining comments.
Marking this one a resolved

File: docs/rpc/openapi.yaml [hunk 11]
```diff
@@ -486,6 +486,93 @@ paths:
...
           If the tip is left unspecified, the stacks chain tip will be selected (only includes confirmed state).
 
+  /v2/clarity/marf/{clarity_marf_key}:
+    post:
+      summary: Get the MARF value for a given key
+      tags:
+        - Smart Contracts
+      operationId: get_clarity_marf_value
+      description: |
+        Attempt to fetch the value of a MARF key. The key is identified with [Clarity MARF Key].
```
Comment (obycode) [inline docs/rpc/openapi.yaml:496]: I think you can delete "The key is identified with [Clarity MARF Key]." since that is covered in the parameters.

File: docs/rpc/openapi.yaml [hunk 12]
```diff
@@ -486,6 +486,93 @@ paths:
...
+            known tip (includes unconfirmed state).
+
+  /v2/clarity/metadata/{contract_address}/{contract_name}/{clarity_metadata_key}:
+    post:
+      summary: Get the contract metadata for the metadata key
+      tags:
+        - Smart Contracts
+      operationId: get_clarity_metadata_key
+      description: |
+        Attempt to fetch the metadata of a contract. The contract is identified with [Stacks Address] and [Contract Name] in the URL path. The metadata key is identified with [Clarity Metadata Key].
```
Comment (obycode) [inline docs/rpc/openapi.yaml:536]: The names used should match, e.g. `[Stacks Address]` => `contract_address`.

File: stackslib/src/net/api/getclaritymetadata.rs [hunk 13]
```diff
@@ -0,0 +1,269 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020-2024 Stacks Open Internet Foundation
```
Comment (obycode) [inline stackslib/src/net/api/getclaritymetadata.rs:2]: nit: for new files, I think they should just have:
```
// Copyright (C) 2024 Stacks Open Internet Foundation
```

**Review by obycode — APPROVED** [review]
LGTM

File: stackslib/src/net/api/getclaritymetadata.rs (hunk 8 above)
Comment (jcnelson) [reply stackslib/src/net/api/getclaritymetadata.rs:39]: Yes, because that fully-built regex does not impose boundaries either.  You need to limit the number of characters this regex can match.

Comment (hugocaillard) [reply stackslib/src/net/api/getclaritymetadata.rs]: Ooh right, sorry I missed that.
Fixed in my last commit

File: stackslib/src/net/api/tests/getclaritymetadata.rs [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+                "ST2DS4MSWSGJ3W9FBC6BVT0Y92S345HY8N3T6AV7R.hello-world"
+            )
+            .unwrap()
+        )
+    );
+
+    assert_eq!(&preamble, request.preamble());
+
+    handler.restart();
+    assert!(handler.clarity_metadata_key.is_none());
```
Comment (jcnelson) [inline stackslib/src/net/api/tests/getclaritymetadata.rs:87]: Can you expand this test to cover error cases?  Specifically, can you verify that the user gets back HTTP 400 for invalid metadata key?

File: stackslib/src/net/api/tests/getclaritymetadata.rs [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+    };
+    assert_eq!(resp.data, expected.serialize());
+
+    // data var metadata
+    let response = responses.remove(0);
+    let resp = response.decode_clarity_metadata_response().unwrap();
+    let expected = DataVariableMetadata {
+        value_type: TypeSignature::IntType,
+    };
+    assert_eq!(resp.data, expected.serialize());
```
Comment (jcnelson) [inline stackslib/src/net/api/tests/getclaritymetadata.rs:272]: Can you add test coverage here for HTTP 400 and HTTP 404 error cases?

File: stackslib/src/net/api/tests/getclaritymetadata.rs (hunk 14 above)
Comment (hugocaillard) [reply stackslib/src/net/api/tests/getclaritymetadata.rs:87]: There are two test case for that below with `vm-metadata::2::contract-size` and `vm-metadata::9::contract-invalid-key`

**Review by jcnelson — APPROVED** [review]

File: stackslib/src/net/api/tests/getclaritymetadata.rs (hunk 15 above)
Comment (hugocaillard) [reply stackslib/src/net/api/tests/getclaritymetadata.rs:292]: Done

**Comment by hugocaillard** [comment]
Thanks for the reviews @jcnelson 🙏 
I marked as resolved all of the addressed comment
Re-requested Brice review as well since it was dismissed, but he already approved earlier.

**Review by obycode — DISMISSED** [review]
🎉

**Comment by hugocaillard** [comment]
@jcnelson I'm actually seeing something weird with the last test I added.

If I add a request after this [call](https://github.com/stacks-network/stacks-core/blob/7f3191b6aa697881d9dc088832dea4a4d1497d14/stackslib/src/net/api/tests/getclaritymetadata.rs#L256-L264), or if I add a previous request hitting this [code path](https://github.com/stacks-network/stacks-core/blob/a94abfda195ca653b6344924eaa62fba7bf2e250/stackslib/src/net/api/getclaritymetadata.rs#L121-L141) (such as a request `vm-metadata::9::invalid-contract-metadata-key`, the following request don't work and [this assertion fails](https://github.com/stacks-network/stacks-core/blob/a94abfda195ca653b6344924eaa62fba7bf2e250/stackslib/src/net/api/tests/mod.rs#L1080).

Is it possible that this code path breaks the test runner?
It works just fine in practice when calling this endpoint

**Comment by jcnelson** [comment]
@hugocaillard Can you push a test case so I can take a look?

**Review by jcnelson — CHANGES_REQUESTED** [review]
Placing a hold on this until we get further clarity from @hugocaillard

File: stackslib/src/net/api/tests/getclaritymarfvalue.rs [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+    // vm-account blaance
+    let response = responses.remove(0);
+    debug!(
+        "Response:\n{}\n",
+        std::str::from_utf8(&response.try_serialize().unwrap()).unwrap()
+    );
+
+    let resp = response.decode_data_var_response().unwrap();
+    let balance = Value::try_deserialize_hex(&resp.data[2..], &TypeSignature::IntType, false);
+    assert_eq!(balance, Ok(Value::Int(256_000_000_000)));
```
Comment (hugocaillard) [inline stackslib/src/net/api/tests/getclaritymarfvalue.rs:201]: I'm not sure that 256M is the right expected value 🤔

Comment (obycode) [reply stackslib/src/net/api/tests/getclaritymarfvalue.rs:201]: Everything looks good to me, but I'm also unclear how this balance reaches 256,000 STX.

Comment (hugocaillard) [reply stackslib/src/net/api/tests/getclaritymarfvalue.rs:201]: @jcnelson Does that make sense to you?

**Review by jcnelson — APPROVED** [review]

**Review by obycode — APPROVED** [review]
LGTM!
